src/IO/MySQLBinlogEventReadBuffer.cpp

```diff
@@ -51,7 +51,8 @@ bool MySQLBinlogEventReadBuffer::nextImpl()
         checksum_buff_size = checksum_buff_limit = checksum_signature_length;
     }
 
-    if (in.eof())
+    /// Check also available bytes after eof() to prevent empty working_buffer
+    if (in.eof() || !in.hasPendingData())
```

Comment:
this change looks pointless because https://github.com/ClickHouse/ClickHouse/blob/c556ad96f6631b18613b8463c1bf1f3ab9965a06/src/IO/ReadBuffer.h#L103-L106
and it should be impossible that `hasPendingData() == false` after `next()` returned `true` (otherwise it is a bug in read buffer).

So looks like the fix in `src/IO/MySQLPacketPayloadReadBuffer.cpp` is enough to fix the issue.

Reply:
It is correct, but together with first fix for `ReadBuffer.h`, what would happen if `MySQLBinlogEventReadBuffer` would be used with different buffer, not `MySQLPacketPayloadReadBuffer` ? Also finally `ReadBuffer::next()` has assert to prevent empty buffers from its impl, which requires the impl to be more correct.

Reply:
>  what would happen if MySQLBinlogEventReadBuffer would be used with different buffer, not MySQLPacketPayloadReadBuffer ? 

then if this another buffer returns `true` from `next()`, but in fact `working_buffer.empty()`, then this buffer has a bug and it should be fixed. This check in `MySQLBinlogEventReadBuffer.cpp` will just hide the bug.